File: Data Structures & Algorithms/island-perimeter/submission-0.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        visited = [[False for _ in range(n)] for _ in range(m)]
        directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        res = 0

        def dfs(i, j):
            nonlocal visited, res
            visited[i][j] = True

            for direction in directions:
                u = i + direction[0]
                v = j + direction[1]
                
                if u < 0 or u >= m or v < 0 or v >= n:
                    res += 1
                    continue
            
                if visited[u][v]:
                    continue
                
                if grid[u][v] != 1:
                    res += 1
                    continue
                
                dfs(u, v)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    dfs(i, j)
                    return res

File: Data Structures & Algorithms/spiral-matrix/submission-0.cpp
class Solution {
public:
    const vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    int curDir = 0;
    vector<vector<bool>> visited;
    vector<int> res;

    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        const int m = matrix.size(); // num rows;
        const int n = matrix.at(0).size(); // num columns;
        visited = vector(m, vector(n, false));
        dfs(matrix, 0, 0, m, n, 0);
        return res;
    }

    void dfs(vector<vector<int>>& matrix, int x, int y, const int m, const int n, int cnt) {
        if (cnt >= 2) {
            return;
        }

        if (cnt == 0) {
            res.push_back(matrix[x][y]);
            visited[x][y] = true;
        }
        
        int u = x + directions[curDir].first;
        int v = y + directions[curDir].second;
        if (u < 0 || u >= m || v < 0 || v >= n || visited[u][v]) {
            curDir = (curDir + 1) % 4;
            dfs(matrix, x, y, m, n, cnt + 1);
        } else {
            dfs(matrix, u, v, m, n, 0);
        }
    }
};
